File: core/player/ships/enemies/manager.js
import { sizeOf } from "../../../utils/constants.js";
import { randomNum } from "../../../utils/random.js";
import { world } from "../../../world/world.js";
import Explosion, { explosions } from "../../prop/explosion.js";
import { destroyedShips } from "../ship.js";
import Bomber from "./bomber.js";
import FleetDrone from "./fleet.js";
import Miner from "./miner.js";
import MissileLaucher from "./missileLaucher.js";
import Sniper from "./sniper.js";
import Tormenter from "./tormenter.js";

export default class EnemyManager {
    static ships = [];
    // static types = [AI];
    static types = [FleetDrone, Tormenter, Sniper, Bomber, Miner, MissileLaucher];

    static init() {
        for (let i = 0; i < sizeOf.ship; i++) {
            EnemyManager.ships.push(new this.types[Math.floor(randomNum(0, this.types.length))]({ x: randomNum(0, world.width), y: randomNum(0, world.height), angle: randomNum(-Math.PI * 2, Math.PI * 2), width: 20, height: 20, color: "red" }));
        }
    }

    static destroy(enemy) {
        const index = EnemyManager.ships.indexOf(enemy);
        
        if (index > -1) {
            enemy.state = "dead";
            explosions.push(new Explosion(enemy.x, enemy.y));
            destroyedShips.push(EnemyManager.ships.splice(index, 1));
        }
    }

    static render() {
        for (const ship of EnemyManager.ships) {
            ship.render();
        }
    }

    static update(t, dt) {
        for (const ship of EnemyManager.ships) {
            ship.update(t, dt);
        }

    }
}

File: void_front/game/core/player/ships/enemies/manager.py
from __future__ import annotations

from game.core.player.prop.explosion import Explosion, explosions
from game.core.player.ships.enemies.bomber import Bomber
from game.core.player.ships.enemies.fleet import FleetDrone
from game.core.player.ships.enemies.miner import Miner
from game.core.player.ships.enemies.missileLaucher import MissileLaucher
from game.core.player.ships.enemies.sniper import Sniper
from game.core.player.ships.enemies.tormenter import Tormenter
from game.core.player.ships.ship import destroyedShips
from game.core.utils.constants import sizeOf
from game.core.world.world import world
import numpy as np


class EnemyManager:
    ships: list = []
    types: list = []

    @staticmethod
    def init(num_enemies: int = sizeOf.ship):
        EnemyManager.types = [
            FleetDrone,
            Bomber,
            Miner,
            MissileLaucher,
            Sniper,
            Tormenter,
        ]
        EnemyManager.ships.clear()
        for _ in range(num_enemies):
            cls = EnemyManager.types[int(np.random.uniform(0, len(EnemyManager.types)))]
            EnemyManager.ships.append(
                cls(
                    x=np.random.uniform(0, world.width),
                    y=np.random.uniform(0, world.height),
                    angle=np.random.uniform(-np.pi * 2, np.pi * 2),
                )
            )

    @staticmethod
    def destroy(enemy):
        index = EnemyManager.ships.index(enemy) if enemy in EnemyManager.ships else -1
        if index > -1:
            enemy.state = "dead"
            explosions.append(Explosion(enemy.x, enemy.y))
            destroyedShips.append(EnemyManager.ships.pop(index))

    @staticmethod
    def render():
        for ship in EnemyManager.ships:
            ship.render()

    @staticmethod
    def update(t, dt):
        for ship in EnemyManager.ships:
            ship.update(t, dt)
